frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -3029,23 +3029,12 @@ considerCompilerGeneratedCandidates(Context* context,
     return;
   }
 
-  // OK, already concrete, store and return
-  if (!tfs->needsInstantiation()) {
-    candidates.push_back(tfs);
-    return;
-  }
-
-  // need to instantiate before storing
-  auto poi = pointOfInstantiationScope(context, inScope, inPoiScope);
-  auto instantiated = doIsCandidateApplicableInstantiating(context,
-                                                           tfs,
-                                                           ci,
-                                                           poi);
-  CHPL_ASSERT(instantiated.success());
-  CHPL_ASSERT(instantiated.candidate()->untyped()->idIsFunction());
-  CHPL_ASSERT(instantiated.candidate()->instantiatedFrom());
-
-  candidates.push_back(instantiated.candidate());
+  // It seems we don't need to instantiate compiler-generated candidates before
+  // storing. Indeed at time of writing, attempting to do so fails as
+  // it relies on the formalDecl's in UntypedFnSignature which are null for
+  // compiler-generated. Anna, 12/14/2023
```

Comment:
I'm not so sure about this.

The call site looks like this

```
  // inject compiler-generated candidates in a manner similar to below
  // (note that any added candidates are already fully instantiated &
  //  the poiInfo from these is not gathered, because such methods should
  //  always be available in any scope that can refer to the type & are
  //  considered part of the custom type)
  considerCompilerGeneratedCandidates(context, ci, inScope, inPoiScope,
                                      candidates);
```

and then the code goes on to add to `candidates` only functions found by name lookup that are type-compatible with the call.

If there are compiler-generated candidates and non-compiler-generated candidates, this will lead to a situation where the disambiguation process has some instantiated functions and some generic functions. I don't think we can expect it to function correctly in that scenario.

I suspect that either:
 1. The compiler-generated candidates are often concrete, and if they are generic, nothing is relying on their types (e.g. the return type); or
 2. We just haven't found a test that combines compiler-generated with user-provided candidates that shows the potential issue here.

Can we look for a different solution to the problem you are facing here?

Reply:
Makes sense. I checked and the "doesn't need instantiation" code path was never hit in the dyno test suite, until my changes to checking for user-defined overloads of operators.

I guess the better solution would be to let those get instantiated, but I don't know how to get around the fact that we can't instantiate a compiler-generated UFS.

Reply:
> I guess the better solution would be to let those get instantiated, but I don't know how to get around the fact that we can't instantiate a compiler-generated UFS.

Well, ultimately, these should be instantiatable based upon the concrete type we are working with. However, there might be some code that needs to change to enable that.

But anyway it'd be good to have an example going while trying to improve this. Do you have an example that you are working with?

Reply:
I ran into it with checking assignability of `GenericRecord` in my added test.
Here's an equivalent standalone code which segfaults in `testInteractive` (with my skipping-instantiation changes removed).

```
record R {
  var x;
}

var x = __primitive("is assignable type", R);
```

The segfault is hit in `instantiateSignature` for the compiler-generated candidate `=`, since we try to resolve the type of a formal which is `nullptr`.

Reply:
I realized that of course this doesn't work because there's no way to instantiate it -- it's generic and has no default.

I double-checked the prod test cases I thought I was mimicking and it provides either a type parameter or a default. When I tried this same case in the production compiler, it errored out trying to resolve the `init=`.

So, I've changed the implementation in Dyno to error immediately when given a fully generic composite type.

Reply:
I suppose that, on the theory that these primitives should not error but instead only answer a question, the possible answer could be "Can't say, because the type you are asking about is generic". Perhaps conveying that is practical if we have something suggested in https://github.com/chapel-lang/chapel/pull/24058#discussion_r1427187882 where the query is returning a custom struct.

Reply:
(Another approach would be to say that the answer, for a generic type, to all these questions like "is it assignable", is "no")

Reply:
> on the theory that these primitives should not error but instead only answer a question, the possible answer could be "Can't say, because the type you are asking about is generic"

I was thinking about this, but how would I translate that answer out to the eventual return type of the primitive called (bool)? I know returning an empty `QualifiedType` is an option, though that doesn't give an explanation.

> (Another approach would be to say that the answer, for a generic type, to all these questions like "is it assignable", is "no")

That makes the most sense to me logically, but it's not the behavior of prod. I would be ok changing the behavior of prod to match this though, since currently it's an error down the line somewhere trying to instaniate a generic type.

Reply:
> > on the theory that these primitives should not error but instead only answer a question, the possible answer could be "Can't say, because the type you are asking about is generic"

> I was thinking about this, but how would I translate that answer out to the eventual return type of the primitive called (bool)? I know returning an empty QualifiedType is an option, though that doesn't give an explanation.

Yeah, good point. It seems that we'd have to have error checking outside of the query (possibly outside of the primitive); and in that case, the query might as well return `false` in this case.


> > (Another approach would be to say that the answer, for a generic type, to all these questions like "is it assignable", is "no")

> That makes the most sense to me logically, but it's not the behavior of prod. I would be ok changing the behavior of prod to match this though, since currently it's an error down the line somewhere trying to instaniate a generic type.

Yeah, I lean towards this option, given the alternatives.